File: 算法/代码随想录算法PDF/00调试源代码/332.重新安排行程.py
#
# @lc app=leetcode.cn id=332 lang=python3
#
# [332] 重新安排行程
#

# @lc code=start
import collections
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        res = []
        targets = collections.defaultdict(list)
        for ticket in tickets:  # 填充映射
            targets[ticket[0]].append(ticket[1])
        for value in targets.values():  # 排序映射
            value.sort()
        res.append("JFK")

        def backtracking():
            if len(res) == len(tickets) + 1: return True
            for _ in targets[res[-1]]:
                tmp = targets[res[-1]].pop(0)
                res.append(tmp)
                if backtracking(): return True
                res.pop()
                targets[res[-1]].append(tmp)
            return False

        backtracking()
        return res
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/332.重新安排行程.cpp
/*
 * @lc app=leetcode.cn id=332 lang=cpp
 *
 * [332] 重新安排行程
 */

// @lc code=start
class Solution {
public:
    unordered_map<string, deque<string>> targets;
    vector<string> res;
    int ticketNum;
    bool backtracking(const string& start) {
        if (res.size() == ticketNum + 1) return true;
        for (int i = 0; i < targets[start].size(); i++) {
            string tmp = targets[start].front();
            targets[start].pop_front();
            res.push_back(tmp);
            if (backtracking(tmp)) return true;
            res.pop_back();
            targets[start].push_back(tmp);
        }
        return false;
    }
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        targets.clear();
        res.clear();
        for (const vector<string>& ticket : tickets) {  // 初始化目标映射
            targets[ticket[0]].push_back(ticket[1]);
        }
        res.push_back("JFK");
        ticketNum = tickets.size();
        for (pair<const string, deque<string>>& target : targets) {
            sort(target.second.begin(), target.second.end());
        }
        backtracking("JFK");
        return res;
    }
};
// @lc code=end

